File: packages/contracts/src/gate-result.ts
import { z } from 'zod';

export const GateVerdictSchema = z.enum(['pass', 'fail', 'pending', 'skipped']);

export const GateResultCreateInputSchema = z.object({
  promotionRunId: z.string(),
  stageId: z.string(),
  verdict: GateVerdictSchema,
  metricType: z.string(),
  observedValue: z.number().optional(),
  threshold: z.number(),
  metadata: z.record(z.string(), z.unknown()),
});

export type GateVerdict = z.infer<typeof GateVerdictSchema>;
export type GateResultCreateInput = z.infer<typeof GateResultCreateInputSchema>;

File: kotlin/modules/contracts/bin/main/com/ffpromo/contracts/GateResult.kt
package com.ffpromo.contracts

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class GateResultCreateInput(
    val promotionRunId: String,
    val stageId: String,
    val verdict: GateVerdict,
    val metricType: String,
    val observedValue: Double? = null,
    val threshold: Double,
    val metadata: Map<String, JsonElement>,
)

/** Worker/repository alias matching plan terminology. */
typealias GateResultRecord = GateResultCreateInput
